File: apps/connect/modules/reflections-audio/android/src/main/java/expo/modules/reflectionsaudio/ReflectionsAudioModule.kt
package expo.modules.reflectionsaudio

import android.content.Context
import android.media.AudioDeviceCallback
import android.media.AudioDeviceInfo
import android.media.AudioManager
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ReflectionsAudioModule : Module() {
  private var deviceCallback: AudioDeviceCallback? = null

  private val audioManager: AudioManager?
    get() = appContext.reactContext?.getSystemService(Context.AUDIO_SERVICE) as? AudioManager

  override fun definition() = ModuleDefinition {
    // Referenced from JS via `requireNativeModule('ReflectionsAudio')`.
    Name("ReflectionsAudio")

    Events("onAudioRouteChange")

    // Android intentionally does NOT enable hardware AEC for reaction recording. Enabling it would
    // require patching expo-camera's audio source (VOICE_COMMUNICATION), which is a maintenance
    // liability with device-variable results. Echo is handled by the JS volume policy instead.
    // Kept as a no-op so JS can call the same audio-session API on both platforms.
    AsyncFunction("setVoiceChatModeAsync") {
      // no-op
    }

    AsyncFunction("beginVoiceChatGuardAsync") {
      // no-op
    }

    AsyncFunction("reassertVoiceChatModeAsync") {
      // no-op
    }

    AsyncFunction("endVoiceChatGuardAsync") {
      // no-op
    }

    AsyncFunction("prepareParentRecordingPlaybackAsync") { _: String, _: Double, _: Double ->
      // no-op — iOS only; Android uses expo-av with muted speaker policy.
    }

    AsyncFunction("startParentRecordingPlaybackAsync") { _: String, _: Double, _: Double ->
      // no-op — parent audio stays on expo-av during recording on Android.
    }

    AsyncFunction("stopParentRecordingPlaybackAsync") {
      // no-op
    }

    AsyncFunction("setPlaybackModeAsync") {
      // no-op
    }

    Function("getAudioRoute") {
      describeCurrentRoute()
    }

    Function("getAudioSessionInfo") {
      mapOf(
        "category" to "android",
        "mode" to "none",
        "isPlayAndRecord" to false,
        "isVoiceChatMode" to false,
        "isOtherAudioPlaying" to false,
        "voiceChatGuardActive" to false,
        "nativeParentPlaybackActive" to false,
        "nativeParentPlaying" to false,
        "nativeParentVolume" to 0,
        "nativeParentRate" to 0,
        "nativeParentTimeSec" to 0,
        "nativeModuleLoaded" to true,
        "nativeModuleVersion" to 3,
        "outputs" to describeCurrentRoute()["outputs"],
        "inputs" to emptyList<String>(),
        "hasHeadphones" to (describeCurrentRoute()["hasHeadphones"] == true),
      )
    }

    OnStartObserving {
      registerDeviceCallback()
    }

    OnStopObserving {
      unregisterDeviceCallback()
    }
  }

  private fun describeCurrentRoute(): Map<String, Any> {
    val manager = audioManager
      ?: return mapOf("hasHeadphones" to false, "outputs" to emptyList<String>())

    val devices = manager.getDevices(AudioManager.GET_DEVICES_OUTPUTS)
    val headphoneTypes = setOf(
      AudioDeviceInfo.TYPE_WIRED_HEADPHONES,
      AudioDeviceInfo.TYPE_WIRED_HEADSET,
      AudioDeviceInfo.TYPE_BLUETOOTH_A2DP,
      AudioDeviceInfo.TYPE_BLUETOOTH_SCO,
      AudioDeviceInfo.TYPE_USB_HEADSET,
      AudioDeviceInfo.TYPE_HEARING_AID,
    )
    val hasHeadphones = devices.any { headphoneTypes.contains(it.type) }
    return mapOf(
      "hasHeadphones" to hasHeadphones,
      "outputs" to devices.map { it.type.toString() },
    )
  }

  private fun registerDeviceCallback() {
    val manager = audioManager ?: return
    val callback = object : AudioDeviceCallback() {
      override fun onAudioDevicesAdded(addedDevices: Array<out AudioDeviceInfo>?) {
        sendEvent("onAudioRouteChange", describeCurrentRoute())
      }

      override fun onAudioDevicesRemoved(removedDevices: Array<out AudioDeviceInfo>?) {
        sendEvent("onAudioRouteChange", describeCurrentRoute())
      }
    }
    manager.registerAudioDeviceCallback(callback, null)
    deviceCallback = callback
  }

  private fun unregisterDeviceCallback() {
    val manager = audioManager ?: return
    deviceCallback?.let { manager.unregisterAudioDeviceCallback(it) }
    deviceCallback = null
  }
}

File: apps/connect/modules/reflections-audio/src/ReflectionsAudioModule.ts
import { NativeModule, requireNativeModule } from 'expo';

import type { AudioOutputRoute, ReflectionsAudioModuleEvents } from './ReflectionsAudio.types';

export type AudioSessionInfo = {
  category: string;
  mode: string;
  isPlayAndRecord?: boolean;
  isVoiceChatMode?: boolean;
  isOtherAudioPlaying?: boolean;
  voiceChatGuardActive: boolean;
  nativeParentPlaybackActive: boolean;
  nativeParentPlaying?: boolean;
  nativeParentVolume?: number;
  nativeParentRate?: number;
  nativeParentTimeSec?: number;
  nativeModuleLoaded: boolean;
  outputs?: string[];
  inputs?: string[];
  hasHeadphones?: boolean;
};

export declare class ReflectionsAudioModule extends NativeModule<ReflectionsAudioModuleEvents> {
  /**
   * iOS: switches the shared `AVAudioSession` to `PlayAndRecord` + `VoiceChat` mode, which routes
   * capture through the hardware Voice-Processing I/O unit (acoustic echo cancellation). This is
   * what lets a Companion record a selfie reaction while the parent Reflection plays out loud
   * without the playback bleeding into the recording.
   *
   * Android: intentionally a no-op. We deliberately do NOT enable platform AEC on Android (it would
   * require patching expo-camera's audio source); echo is handled by the JS volume policy instead.
   */
  setVoiceChatModeAsync(): Promise<void>;

  /**
   * iOS: enables VoiceChat and keeps re-applying it while expo-camera may clobber the session.
   * Android: no-op.
   */
  beginVoiceChatGuardAsync(): Promise<void>;

  /** iOS: re-applies VoiceChat when the guard is active. Android: no-op. */
  reassertVoiceChatModeAsync(): Promise<void>;

  /** iOS: tears down the guard and native parent playback. Android: no-op. */
  endVoiceChatGuardAsync(): Promise<void>;

  /**
   * iOS: plays parent Reflection audio through the VoiceChat session during selfie recording so
   * hardware AEC receives a reference signal. expo-av bypasses this path and causes mic bleed.
   * Android: no-op (expo-av handles parent audio).
   */
  startParentRecordingPlaybackAsync(
    url: string,
    startMs: number,
    volume: number,
  ): Promise<void>;

  /** Stops native parent playback started by `startParentRecordingPlaybackAsync`. */
  stopParentRecordingPlaybackAsync(): Promise<void>;

  /**
   * iOS: restores a playback-optimised `AVAudioSession` (`Playback` + `MoviePlayback`) after a
   * recording session so subsequent video playback is full fidelity. Android: no-op.
   */
  setPlaybackModeAsync(): Promise<void>;

  /** Returns the device's current audio output route (synchronous). */
  getAudioRoute(): AudioOutputRoute;

  /** Returns the current AVAudioSession category/mode for dev diagnostics. */
  getAudioSessionInfo(): AudioSessionInfo;
}

/**
 * The native module is only present in a development/production build that was compiled with this
 * local module. We resolve it defensively so that Metro web bundling, Expo Go, or an out-of-date
 * dev client degrade gracefully (every method becomes a safe no-op) instead of crashing at import.
 */
let nativeModule: ReflectionsAudioModule | null = null;
try {
  nativeModule = requireNativeModule<ReflectionsAudioModule>('ReflectionsAudio');
} catch {
  nativeModule = null;
}

export default nativeModule;
